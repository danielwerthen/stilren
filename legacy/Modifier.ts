function getClassName(instance: Modifier) {
  return Reflect.getPrototypeOf(instance).constructor.name;
}

const modifiers: { [key: string]: Modifier } = {};

export type ModifierType = "prefix" | "suffix";

export class Modifier {
  value: string;
  type: ModifierType;
  constructor(value: string, type: ModifierType) {
    const id = `stil-ren-modifier:${getClassName(this)}+${value}`;
    const mod = modifiers[id];
    this.value = value;
    this.type = type;
    if (mod) {
      return mod;
    }
    modifiers[id] = this;
  }
  canCombine(_withMod: Modifier) {
    return false;
  }
  combine(withMod: Modifier): Modifier {
    throw new Error("Cannot combine with " + getClassName(withMod));
  }
  getKey() {
    return this.value;
  }

  toString() {
    return this.getKey();
  }

  static combineModifiers(mods: Modifier[]) {
    return mods.reduce<Modifier[]>((sum: Modifier[], item: Modifier) => {
      const last = sum[sum.length - 1];
      if (last && last.canCombine(item)) {
        return [...sum.slice(0, sum.length - 1), last.combine(item)];
      }
      return [...sum, item];
    }, []);
  }
}

export class MediaModifier extends Modifier {
  constructor(value: string) {
    super(value, "prefix");
  }
  canCombine(withMod: Modifier) {
    return withMod instanceof MediaModifier;
  }
  combine(withMod: Modifier): Modifier {
    const alpha = this;
    const beta = withMod;
    if (alpha === beta) {
      return beta;
    }
    const left = alpha.value < beta.value ? alpha : beta;
    const right = alpha.value < beta.value ? beta : alpha;
    return new MediaModifier(`${left.value} or ${right.value}`);
  }
  getKey() {
    return `@media ${this.value}`;
  }
}

export class PseudoModifier extends Modifier {
  constructor(value: string) {
    super(value, "suffix");
  }
  canCombine(withMod: Modifier) {
    return withMod instanceof PseudoModifier;
  }
  combine(withMod: Modifier): Modifier {
    const alpha = this;
    const beta = withMod;
    if (alpha === beta) {
      return beta;
    }
    let left: Modifier;
    let right: Modifier;
    const aStart = alpha.value.startsWith(":");
    const bStart = beta.value.startsWith(":");
    if (aStart && bStart) {
      left = alpha.value < beta.value ? alpha : beta;
      right = alpha.value < beta.value ? beta : alpha;
    } else if (aStart) {
      left = beta;
      right = alpha;
    } else if (bStart) {
      left = alpha;
      right = beta;
    } else {
      left = alpha.value < beta.value ? alpha : beta;
      right = alpha.value < beta.value ? beta : alpha;
    }

    return new PseudoModifier(`${left.value}:${right.value}`);
  }
  getKey() {
    return `:${this.value}`;
  }
}
