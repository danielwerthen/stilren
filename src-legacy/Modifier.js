function getClassName(instance) {
  return Reflect.getPrototypeOf(instance).constructor.name;
}

const modifiers = {};

export class Modifier {
  constructor(value) {
    const id = Symbol.for(`stil-ren-modifier:${getClassName(this)}+${value}`);
    const mod = modifiers[id];
    if (mod) {
      return mod;
    }
    this.value = value;
    Object.freeze(this);
    modifiers[id] = this;
  }
  canCombine(withMod) {
    return false;
  }
  combine(withMod) {
    throw new Error('Cannot combine with ' + getClassName(withMod));
  }
  getKey() {
    return this.value;
  }

  toString() {
    return this.getKey();
  }

  static getModifierById(id) {
    return getModifier(id);
  }

  static combineModifiers(mods) {
    return mods.reduce((sum, item) => {
      const last = sum[sum.length - 1];
      if (last && last.canCombine(item)) {
        return [...sum.slice(0, sum.length - 1), last.combine(item)];
      }
      return [...sum, item];
    }, []);
  }
}

export class MediaModifier extends Modifier {
  canCombine(withMod) {
    return withMod instanceof MediaModifier;
  }
  combine(withMod, optional) {
    const alpha = optional ? withMod : this;
    const beta = optional ? optional : withMod;
    if (alpha === beta) {
      return beta;
    }
    const left = alpha.value < beta.value ? alpha : beta;
    const right = alpha.value < beta.value ? beta : alpha;
    return new MediaModifier(`${left.value} and ${right.value}`);
  }
  getKey() {
    return `@media ${this.value}`;
  }
}

export class PseudoModifier extends Modifier {
  canCombine(withMod) {
    return withMod instanceof PseudoModifier;
  }
  combine(withMod, optional) {
    const alpha = optional ? withMod : this;
    const beta = optional ? optional : withMod;
    if (alpha === beta) {
      return beta;
    }
    const left = alpha.value < beta.value ? alpha : beta;
    const right = alpha.value < beta.value ? beta : alpha;
    return new PseudoModifier(`${left.value}:${right.value}`);
  }
  getKey() {
    return `:${this.value}`;
  }
}
