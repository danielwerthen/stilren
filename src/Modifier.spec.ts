import { Modifier, MediaModifier, PseudoModifier } from "./Modifier";

describe("Modifier", () => {
  it("enforces singletons", () => {
    const a = new Modifier("a", "prefix");
    const b = new Modifier("a", "prefix");
    expect(a).toBe(b);
  });
  it("can combine media modifiers", () => {
    const a = new MediaModifier("(min-width: 450px)");
    const b = new MediaModifier("(max-width: 800px)");
    const aAndB = a.canCombine(b) && a.combine(b);
    expect(typeof aAndB !== "boolean" && aAndB.getKey()).toMatchSnapshot();
  });
  it("can not combine pseudo and media modifiers", () => {
    const a = new MediaModifier("(min-width: 450px)");
    const b = new PseudoModifier("hover");
    const aAndB = a.canCombine(b) && a.combine(b);
    expect(aAndB).toEqual(false);
  });
  it("can combine pseudo modifiers", () => {
    const a = new PseudoModifier("hover");
    const b = new PseudoModifier("focus");
    const aAndB = a.canCombine(b) && a.combine(b);
    expect(aAndB instanceof PseudoModifier && aAndB.getKey()).toMatchSnapshot();
  });

  it("can reduce multiple modifiers", () => {
    const a = new MediaModifier("(min-width: 450px)");
    const b = new MediaModifier("(max-width: 800px)");
    const c = new PseudoModifier("hover");
    const d = new PseudoModifier("focus");
    const mods = [d, a, b, c];
    const reduced = PseudoModifier.combineModifiers(mods);
    expect(reduced.length).toEqual(3);
  });
});
