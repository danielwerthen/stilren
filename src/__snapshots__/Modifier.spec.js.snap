// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modifier can combine media modifiers 1`] = `"@media (min-width: 450px) and (max-width: 800px)"`;

exports[`Modifier can combine pseudo modifiers 1`] = `":hover:focus"`;
